File: crates/vleo-core/src/physics/orbit.rs
//! Orbital mechanics.
//!
//! Two-body with the J2 secular terms. Nothing here integrates a trajectory —
//! this is a design tool, and a design tool answers "what does this altitude
//! cost me" rather than "where will it be on Tuesday". The propagator lives
//! behind the campaign face, and it consumes these same relations.

use vleo_units::constants::*;
use vleo_units::pmath;
use vleo_units::*;

/// Geocentric radius from altitude above the WGS-84 equatorial radius.
pub fn radius(altitude: Length) -> Length {
    Length::new(R_EARTH.get() + altitude.get())
}

/// Altitude from geocentric radius — the inverse, written once so no node does
/// the subtraction with a different Earth radius.
pub fn altitude(radius: Length) -> Length {
    Length::new(radius.get() - R_EARTH.get())
}

/// Circular orbital speed, `sqrt(mu/r)`.
///
/// At 200 km this is 7.784 km/s and at 400 km it is 7.669 km/s. The difference
/// looks small and is not: drag goes as the square of it, and density between
/// those two altitudes changes by a factor of several hundred.
pub fn circular_velocity(radius: Length) -> Velocity {
    Velocity::new(pmath::sqrt(MU_EARTH / radius.get()))
}

/// Speed on an elliptical orbit at radius `r`, from the vis-viva equation.
pub fn vis_viva(radius: Length, semi_major_axis: Length) -> Velocity {
    Velocity::new(pmath::sqrt(
        MU_EARTH * (2.0 / radius.get() - 1.0 / semi_major_axis.get()),
    ))
}

/// Orbital period, `2·pi·sqrt(a^3/mu)`.
pub fn period(semi_major_axis: Length) -> Time {
    let a = semi_major_axis.get();
    Time::new(pmath::TAU * pmath::sqrt(a * a * a / MU_EARTH))
}

/// Mean motion, `sqrt(mu/a^3)`.
pub fn mean_motion(semi_major_axis: Length) -> AngularRate {
    let a = semi_major_axis.get();
    AngularRate::new(pmath::sqrt(MU_EARTH / (a * a * a)))
}

/// Apoapsis and periapsis radii from `a` and `e`.
pub fn apsides(semi_major_axis: Length, eccentricity: f64) -> (Length, Length) {
    let a = semi_major_axis.get();
    (
        Length::new(a * (1.0 + eccentricity)),
        Length::new(a * (1.0 - eccentricity)),
    )
}

/// Speed of the sub-satellite point over the ground.
///
/// Not the orbital speed: the ground track moves more slowly, in the ratio of
/// the Earth's radius to the orbital radius, and the difference is what sets
/// how long a target stays in a sensor's field of view.
pub fn ground_track_speed(orbital_speed: Velocity, radius: Length) -> Velocity {
    Velocity::new(orbital_speed.get() * R_EARTH.get() / radius.get())
}

/// Secular rate of change of right ascension of the ascending node, from the
/// second zonal harmonic.
///
/// ```text
/// dOmega/dt = -1.5 · J2 · sqrt(mu) · Re^2 · cos(i) / ((1-e^2)^2 · a^(7/2))
/// ```
///
/// Negative for a prograde orbit — the node regresses. This is the relation a
/// Sun-synchronous orbit is designed against, and the one that decides whether
/// a constellation's planes stay where they were put.
pub fn nodal_regression(semi_major_axis: Length, eccentricity: f64, inclination: Angle) -> AngularRate {
    let a = semi_major_axis.get();
    let p = (1.0 - eccentricity * eccentricity) * (1.0 - eccentricity * eccentricity);
    let re = R_EARTH.get();
    AngularRate::new(
        -1.5 * J2_EARTH * pmath::sqrt(MU_EARTH) * re * re * inclination.cos()
            / (p * pmath::powf(a, 3.5)),
    )
}

/// Mean rate at which the Sun's right ascension advances, one revolution per
/// tropical year. A Sun-synchronous orbit is one whose nodal regression matches
/// it.
pub const SUN_RA_RATE: AngularRate = AngularRate::new(1.991_063_853e-7);

/// The inclination at which nodal regression equals the Sun's apparent motion.
///
/// Returns `None` when no such inclination exists — which happens above roughly
/// 5 970 km, and is a real answer rather than a failure: the caller raises the
/// fault, because only the caller knows whether being asked was reasonable.
pub fn sun_synchronous_inclination(semi_major_axis: Length, eccentricity: f64) -> Option<Angle> {
    let a = semi_major_axis.get();
    let p = (1.0 - eccentricity * eccentricity) * (1.0 - eccentricity * eccentricity);
    let re = R_EARTH.get();
    let cos_i = -SUN_RA_RATE.get() * p * pmath::powf(a, 3.5)
        / (1.5 * J2_EARTH * pmath::sqrt(MU_EARTH) * re * re);
    if cos_i < -1.0 || cos_i > 1.0 {
        return None;
    }
    Some(Angle::new(pmath::acos(cos_i)))
}

/// Half-angle of the Earth as seen from orbit, `asin(Re/r)`.
pub fn earth_angular_radius(radius: Length) -> Angle {
    Angle::new(pmath::asin(R_EARTH.get() / radius.get()))
}

/// Earth-central angle to the horizon at a given minimum elevation.
///
/// The geometry every access, swath and contact-time calculation rests on:
///
/// ```text
/// lambda = acos( (Re/r)·cos(eps) ) - eps
/// ```
pub fn earth_central_angle(radius: Length, min_elevation: Angle) -> Angle {
    let ratio = R_EARTH.get() / radius.get();
    Angle::new(pmath::acos(ratio * min_elevation.cos()) - min_elevation.get())
}

/// Slant range to a point at the edge of the access circle.
pub fn slant_range(radius: Length, min_elevation: Angle) -> Length {
    let lambda = earth_central_angle(radius, min_elevation);
    let re = R_EARTH.get();
    let r = radius.get();
    Length::new(pmath::sqrt(
        re * re + r * r - 2.0 * re * r * lambda.cos(),
    ))
}

/// Ground swath width for a given Earth-central half-angle.
pub fn swath_width(earth_central_angle: Angle) -> Length {
    Length::new(2.0 * earth_central_angle.get() * R_EARTH_MEAN.get())
}

/// Fraction of an orbit spent in the Earth's shadow, cylindrical model.
///
/// `beta` is the angle between the orbit plane and the Sun direction. Above the
/// critical beta the orbit is fully sunlit and the fraction is zero — which is
/// the case a thermal design must not be sized on, and the case a power design
/// must not be sized on either, for opposite reasons.
pub fn eclipse_fraction(radius: Length, beta: Angle) -> Ratio {
    let re = R_EARTH.get();
    let r = radius.get();
    let cb = pmath::abs(beta.cos());
    // Below this beta the shadow cylinder is missed entirely.
    let beta_star = pmath::asin(re / r);
    if pmath::abs(beta.get()) >= pmath::FRAC_PI_2 - beta_star + 1.0e-12 && cb * r <= re {
        // Fully shadowed geometry does not arise for a circular LEO orbit.
    }
    let arg = pmath::sqrt(r * r - re * re) / (r * cb);
    if arg >= 1.0 {
        return Ratio::new(0.0); // full-sun orbit
    }
    Ratio::new(pmath::acos(arg) / pmath::PI)
}

/// Rate of change of semi-major axis under drag, for a near-circular orbit.
///
/// ```text
/// da/dt = -rho · (Cd·A/m) · sqrt(mu·a)
/// ```
///
/// Negative always. This is the number that decides whether a VLEO mission is a
/// mission or an experiment: at 200 km an uncompensated satellite of ordinary
/// ballistic coefficient loses its orbit in days.
pub fn decay_rate(
    density: MassDensity,
    ballistic_coefficient: f64,
    semi_major_axis: Length,
) -> Velocity {
    // Ballistic coefficient here is m/(Cd·A) in kg/m^2, so its reciprocal is
    // the area-to-mass term the relation wants.
    Velocity::new(
        -density.get() * pmath::sqrt(MU_EARTH * semi_major_axis.get()) / ballistic_coefficient,
    )
}

/// Orbital lifetime estimate, by integrating the decay rate at constant
/// atmospheric conditions.
///
/// Reported as an order of magnitude, never as a date. It assumes the density
/// at each altitude does not change while the satellite passes through it,
/// which over a solar cycle is wrong by a factor of several — and that is the
/// honest accuracy of any lifetime number that is not a Monte Carlo over solar
/// activity.
pub fn lifetime_estimate<F>(
    start_altitude: Length,
    end_altitude: Length,
    ballistic_coefficient: f64,
    steps: u32,
    density_at: F,
) -> Time
where
    F: Fn(Length) -> MassDensity,
{
    let h0 = start_altitude.get();
    let h1 = end_altitude.get();
    let dh = (h1 - h0) / steps as f64;
    let mut t = 0.0;
    for i in 0..steps {
        let h = h0 + dh * (i as f64 + 0.5);
        let a = radius(Length::new(h));
        let rho = density_at(Length::new(h)).get();
        let da_dt = -rho * pmath::sqrt(MU_EARTH * a.get()) / ballistic_coefficient;
        if da_dt >= 0.0 {
            continue;
        }
        t += dh / da_dt;
    }
    Time::new(pmath::abs(t))
}

/// Delta-v needed to hold an altitude against a drag acceleration for a
/// duration. Exact for a low-thrust, continuously-compensated orbit: the drag
/// impulse is the impulse that has to be replaced.
pub fn drag_makeup_delta_v(drag_acceleration: Acceleration, duration: Time) -> Velocity {
    drag_acceleration * duration
}

/// Delta-v for a Hohmann transfer between two circular orbits.
pub fn hohmann_delta_v(r1: Length, r2: Length) -> Velocity {
    let (r1, r2) = (r1.get(), r2.get());
    let a_t = 0.5 * (r1 + r2);
    let v1 = pmath::sqrt(MU_EARTH / r1);
    let v2 = pmath::sqrt(MU_EARTH / r2);
    let vp = pmath::sqrt(MU_EARTH * (2.0 / r1 - 1.0 / a_t));
    let va = pmath::sqrt(MU_EARTH * (2.0 / r2 - 1.0 / a_t));
    Velocity::new(pmath::abs(vp - v1) + pmath::abs(v2 - va))
}

/// Delta-v for a pure plane change at a given speed.
pub fn plane_change_delta_v(speed: Velocity, angle: Angle) -> Velocity {
    Velocity::new(2.0 * speed.get() * pmath::sin(0.5 * angle.get()))
}

/// Delta-v to lower periapsis to a target re-entry altitude — the disposal
/// manoeuvre ISO 24113 requires be budgeted at design time rather than found
/// at the end.
pub fn deorbit_delta_v(from_altitude: Length, target_perigee_altitude: Length) -> Velocity {
    let r1 = radius(from_altitude);
    let r2 = radius(target_perigee_altitude);
    let a_t = Length::new(0.5 * (r1.get() + r2.get()));
    let v_circ = circular_velocity(r1);
    let v_transfer = vis_viva(r1, a_t);
    Velocity::new(pmath::abs(v_circ.get() - v_transfer.get()))
}

/// Rocket equation, solved for propellant mass.
pub fn propellant_mass(dry_mass: Mass, delta_v: Velocity, exhaust_velocity: Velocity) -> Mass {
    Mass::new(dry_mass.get() * (pmath::exp(delta_v.get() / exhaust_velocity.get()) - 1.0))
}

/// Exhaust velocity from specific impulse, `v_e = Isp·g0`.
pub fn exhaust_velocity(specific_impulse: Time) -> Velocity {
    Velocity::new(specific_impulse.get() * G0.get())
}

/// Specific impulse from exhaust velocity.
pub fn specific_impulse(exhaust_velocity: Velocity) -> Time {
    Time::new(exhaust_velocity.get() / G0.get())
}

File: tools/nodes/orbit.py
#!/usr/bin/env python3
"""Orbit and geometry.

One subsystem's rows, and nothing else. Split out of the seeder so that
changing what this subsystem declares does not mean opening the file that
declares every other one — the same reason each subsystem has its own crate.

The helpers come from `seed_helpers`, which owns the shared lists these
append to. Import order is authoring order, so `seed_tree` decides it and this
file only says what the rows are.
"""

from seed_helpers import *          # noqa: F401,F403  the row helpers: D, C, layer, S, unnamed

# =============================================================================
# ORBIT AND GEOMETRY
# =============================================================================
D("orbit_altitude", "Orbit altitude", "orbit", "orbit_geometry", "Length", "Kilometre", "h",
  250.0, 150.0, 450.0,
  "below 150 km the flow stops being free-molecular for a body of this size and every aerodynamic node in the tree is outside its envelope",
  "above 450 km there is not enough air for the air-breathing concept to be the reason for the design",
  "At what altitude does the constellation fly? This is the one parameter the whole architecture is shaped around.",
  "orbitt_case_c1", "systems", "A. Rai / 2026-09-01",
  kpis=["kpi_revisit", "kpi_gsd", "kpi_service_lifetime"],
  note="Move this one value and 54 other nodes become stale. That is the requirement the tool exists to satisfy.")

D("orbit_eccentricity", "Orbit eccentricity", "orbit", "orbit_geometry", "Ratio", "One", "e",
  0.0, 0.0, 0.05,
  "a negative eccentricity is not an orbit",
  "above 0.05 the perigee of a 250 km orbit is below 200 km, where the drag per revolution stops being a small perturbation and the circular-orbit relations used throughout this tree no longer apply",
  "How circular is the orbit?",
  "orbitt_case_c1", "systems", "A. Rai / 2026-09-01",
  note="Moving this reaches 3 nodes. Moving the altitude reaches 54. The difference is the answer to how much of the design a change just invalidated.")

D("orbit_inclination", "Orbit inclination", "orbit", "orbit_geometry", "Angle", "Degree", "i",
  96.6, 0.0, 180.0,
  "inclination is defined on 0..180 degrees",
  "inclination is defined on 0..180 degrees",
  "What is the angle between the orbit plane and the equator?",
  "orbitt_case_c1", "systems", "A. Rai / 2026-09-01")

D("orbit_min_elevation", "Minimum useful elevation angle", "orbit", "orbit_geometry", "Angle", "Degree", "eps",
  10.0, 0.0, 90.0,
  "at the horizon the slant range is longest, the atmosphere is thickest and no useful link or image is obtained; zero is the limit, not a working value",
  "90 degrees is the zenith, at which the access circle has no area",
  "Below what elevation is a target or a ground station no longer usable?",
  "larson_wertz", "systems", "A. Rai / 2026-09-01")

D("orbit_beta_angle", "Beta angle", "orbit", "orbit_geometry", "Angle", "Degree", "beta",
  30.0, -90.0, 90.0,
  "the angle between the orbit plane and the Sun is defined on -90..90",
  "the angle between the orbit plane and the Sun is defined on -90..90",
  "What angle does the Sun make with the orbit plane? It decides eclipse duration and therefore both the power and the thermal design.",
  "larson_wertz", "systems", "A. Rai / 2026-09-01")

D("orbit_mission_duration", "Mission duration", "orbit", "orbit_geometry", "Time", "Year", "T_mis",
  5.0, 0.5, 15.0,
  "below six months the programme cannot amortise a satellite, so it is not the mission being designed",
  "above 15 years the cost model, the degradation model and the battery cycle model are all extrapolated well past their fits",
  "How long must one satellite operate?",
  "orbitt_case_c1", "systems", "A. Rai / 2026-09-01",
  kpis=["kpi_cost_per_year"])

C("orbit_radius", "Geocentric radius", "orbit", "orbit_geometry", "Length", "Metre", "r",
  "How far is the spacecraft from the centre of the Earth?",
  "r = R_earth + h",
  "wgs84", "systems",
  [("h", "orbit_altitude")],
  [("add the WGS-84 equatorial radius to the altitude", "r", "Length", "orbit::radius(h)")],
  6.4e6, 7.0e6,
  "below the Earth's mean radius is not an orbit",
  "above 7000 km is outside the band this tool is scoped to",
  tier="A",
  fixtures=[("250 km", {"h": 250000.0}, 6628137.0, 1e-12, "independent-derivation", "wgs84")])

C("orbit_velocity", "Circular orbital speed", "orbit", "orbit_geometry", "Velocity", "MetrePerSecond", "V",
  "How fast is the spacecraft moving through the atmosphere?",
  "V = sqrt(mu/r)",
  "vallado2013", "systems",
  [("r", "orbit_radius")],
  [("apply the circular two-body speed relation", "v", "Velocity", "orbit::circular_velocity(r)")],
  7000.0, 8200.0,
  "below 7 km/s corresponds to an orbit above 1000 km, outside this tool's band",
  "8.2 km/s exceeds the surface circular speed and is not reachable",
  tier="A",
  assumptions=[("Two-body, circular",
                "at e = 0.05 the speed varies by 5% around the orbit and a drag figure computed at the mean underestimates the perigee pass")],
  fixtures=[("250 km circular", {"r": 6628137.0}, 7754.84549737, 1e-9, "independent-derivation", "vallado2013"),
            ("400 km circular", {"r": 6778137.0}, 7668.55817541, 1e-9, "independent-derivation", "vallado2013")])

C("orbit_period", "Orbital period", "orbit", "orbit_geometry", "Time", "Minute", "T_orb",
  "How long does one revolution take?",
  "T = 2*pi*sqrt(a^3/mu)",
  "vallado2013", "systems",
  [("r", "orbit_radius")],
  [("apply Kepler's third law at the circular radius", "t", "Time", "orbit::period(r)")],
  83.0, 104.0,
  "a period below 5000 s corresponds to an orbit inside the Earth",
  "6200 s corresponds to about 1000 km, outside this tool's band",
  tier="A",
  fixtures=[("250 km circular", {"r": 6628137.0}, 5370.29564631, 1e-9, "independent-derivation", "vallado2013")])

C("orbit_ground_track_speed", "Ground track speed", "orbit", "orbit_geometry", "Velocity", "MetrePerSecond", "V_g",
  "How fast does the sub-satellite point move over the ground?",
  "V_g = V*Re/r",
  "larson_wertz", "systems",
  [("v", "orbit_velocity"), ("r", "orbit_radius")],
  [("scale the orbital speed by the ratio of Earth radius to orbital radius",
    "vg", "Velocity", "orbit::ground_track_speed(v, r)")],
  6000.0, 8000.0,
  "below 6 km/s is outside this tool's altitude band",
  "the ground track can never move faster than the orbital speed",
  tier="A",
  note="Not the orbital speed. The difference is what sets sensor dwell time and therefore signal-to-noise.")

C("orbit_nodal_regression", "Nodal regression rate", "orbit", "orbit_geometry", "AngularRate", "DegreePerDay", "dOmega",
  "How fast does the orbit plane rotate under the second zonal harmonic?",
  "dOmega/dt = -1.5*J2*sqrt(mu)*Re^2*cos(i)/((1-e^2)^2*a^(7/2))",
  "vallado2013", "systems",
  [("r", "orbit_radius"), ("e", "orbit_eccentricity"), ("i", "orbit_inclination")],
  [("evaluate the secular J2 node rate", "d", "AngularRate", "orbit::nodal_regression(r, e.get(), i)")],
  -15.0, 15.0,
  "a regression faster than 15 degrees per day does not occur in this altitude band",
  "a regression faster than 15 degrees per day does not occur in this altitude band",
  tier="A",
  fixtures=[("250 km, 96.6 deg", {"r": 6628137.0, "e": 0.0, "i": 1.68598806}, 2.02216606e-7, 1e-6, "independent-derivation", "vallado2013")])

C("orbit_sun_sync_inclination", "Sun-synchronous inclination", "orbit", "orbit_geometry", "Angle", "Degree", "i_ss",
  "What inclination makes the orbit plane keep pace with the Sun?",
  "cos(i) = -Omega_sun*(1-e^2)^2*a^(7/2)/(1.5*J2*sqrt(mu)*Re^2)",
  "vallado2013", "systems",
  [("r", "orbit_radius"), ("e", "orbit_eccentricity")],
  [("solve the nodal regression relation for the inclination that matches the Sun's apparent motion; refuse when no such inclination exists",
    "i", "Angle",
    "match orbit::sun_synchronous_inclination(r, e.get()) { Some(i) => i, None => return Err(Fault::Degenerate { node: NODE_ID, field: \"r\", reason: \"no inclination gives Sun-synchronous regression at this radius\" }) }")],
  85.0, 115.0,
  "a Sun-synchronous inclination is always retrograde and above about 95 degrees in this band",
  "above 115 degrees no Sun-synchronous solution exists below 2000 km",
  tier="A",
  fixtures=[("250 km circular", {"r": 6628137.0, "e": 0.0}, 1.68420864, 1e-6, "independent-derivation", "vallado2013")])

C("orbit_earth_central_angle", "Earth central half-angle", "orbit", "orbit_geometry", "Angle", "Degree", "lambda",
  "How much of the Earth's surface is in view at the working elevation?",
  "lambda = acos((Re/r)*cos(eps)) - eps",
  "larson_wertz", "systems",
  [("r", "orbit_radius"), ("eps", "orbit_min_elevation")],
  [("apply the spherical Earth access geometry", "l", "Angle", "orbit::earth_central_angle(r, eps)")],
  0.0, 86.0,
  "a negative half-angle means the target is below the horizon",
  "beyond 85 degrees the whole visible hemisphere is in view, which does not occur in this band",
  tier="A",
  fixtures=[("250 km, 10 deg elevation", {"r": 6628137.0, "eps": 0.17453293}, 0.150429309, 1e-6, "independent-derivation", "larson_wertz")])

C("orbit_swath_width", "Ground swath width", "orbit", "orbit_geometry", "Length", "Kilometre", "W",
  "How wide a strip of ground does one pass cover?",
  "W = 2*lambda*Re",
  "larson_wertz", "systems",
  [("lam", "orbit_earth_central_angle")],
  [("convert the Earth-central half-angle into a great-circle width", "w", "Length", "orbit::swath_width(lam)")],
  1.0, 40000.0,
  "a swath below one kilometre would not be a swath",
  "a swath cannot exceed the Earth's circumference",
  tier="A",
  kpis=["kpi_revisit"])

C("orbit_slant_range", "Slant range at working elevation", "orbit", "orbit_geometry", "Length", "Kilometre", "d",
  "How far away is a target or ground station at the edge of the access circle?",
  "d = sqrt(Re^2 + r^2 - 2*Re*r*cos(lambda))",
  "larson_wertz", "systems",
  [("r", "orbit_radius"), ("eps", "orbit_min_elevation")],
  [("apply the law of cosines across the access triangle", "d", "Length", "orbit::slant_range(r, eps)")],
  100.0, 4000.0,
  "the slant range is at least the altitude",
  "beyond 4000 km the geometry is outside this tool's band",
  tier="A")

C("orbit_eclipse_fraction", "Eclipse fraction of the orbit", "orbit", "orbit_geometry", "Ratio", "One", "f_ecl",
  "What fraction of each revolution is spent in the Earth's shadow?",
  "f_ecl = (1/pi)*acos(sqrt(r^2 - Re^2)/(r*cos(beta)))",
  "larson_wertz", "systems",
  [("r", "orbit_radius"), ("beta", "orbit_beta_angle")],
  [("apply the cylindrical shadow model; a full-sun orbit returns zero rather than a fault",
    "f", "Ratio", "orbit::eclipse_fraction(r, beta)")],
  0.0, 0.45,
  "a negative eclipse fraction is not a fraction",
  "no circular orbit in this band spends more than 45% of a revolution in shadow",
  tier="A",
  assumptions=[("Cylindrical shadow, no penumbra",
                "the penumbra adds about 10 seconds per orbit, which matters for a precise thermal transient and not for a power budget")],
  fixtures=[("250 km, beta = 30 deg", {"r": 6628137.0, "beta": 0.52359878}, 0.398283597, 1e-6, "independent-derivation", "larson_wertz")])

C("orbit_decay_rate", "Orbital decay rate", "orbit", "orbit_geometry", "Velocity", "MetrePerSecond", "da_dt",
  "How fast does the orbit fall if nothing compensates the drag?",
  "da/dt = -rho*sqrt(mu*a)/BC",
  "vallado2013", "systems",
  [("rho", "env_mass_density"), ("bc", "aero_ballistic_coefficient"), ("r", "orbit_radius")],
  [("apply the circular-orbit secular decay relation", "d", "Velocity", "orbit::decay_rate(rho, bc.get(), r)")],
  -10.0, 0.0,
  "a decay faster than 10 m/s of semi-major axis per second is re-entry, not an orbit",
  "drag cannot raise an orbit; a positive value here is a sign error upstream",
  tier="B",
  kpis=["kpi_service_lifetime"],
  view=("line", {"over": "orbit_altitude", "points": 60}))

C("orbit_makeup_delta_v", "Drag make-up delta-v per year", "orbit", "orbit_geometry", "Velocity", "MetrePerSecond", "dv_dm",
  "How much delta-v a year does holding this altitude cost?",
  "dv = (D/m)*t",
  "vallado2013", "systems",
  [("a_drag", "aero_drag_acceleration")],
  [("integrate the drag deceleration over one Julian year",
    "dv", "Velocity", "orbit::drag_makeup_delta_v(a_drag, Time::new(31_557_600.0))")],
  0.0, 100000.0,
  "make-up delta-v cannot be negative",
  "above 100 km/s a year no propulsion system of any kind closes, and the answer is that the altitude is wrong",
  tier="B",
  note="This is the number an air-breathing system exists to make free. A stored-propellant design at 250 km spends several kilometres per second a year on it.")

C("orbit_deorbit_delta_v", "Disposal delta-v", "orbit", "orbit_geometry", "Velocity", "MetrePerSecond", "dv_dis",
  "What does the end-of-life disposal manoeuvre cost, as required by the debris mitigation standard?",
  "dv = |V_circ - V_transfer(r, a_t)|",
  "iso24113", "systems",
  [("h", "orbit_altitude")],
  [("lower perigee to a 60 km re-entry interface on a Hohmann transfer",
    "dv", "Velocity", "orbit::deorbit_delta_v(h, Length::from_km(60.0))")],
  0.0, 500.0,
  "a disposal manoeuvre cannot cost negative delta-v",
  "above 500 m/s the disposal budget dominates the design and the altitude choice should be revisited",
  tier="A",
  note="Carried as a design variable rather than discovered at the end, because ISO 24113 makes it a requirement and not a courtesy.")

C("orbit_lifetime_uncontrolled", "Uncontrolled orbital lifetime", "orbit", "orbit_geometry", "Time", "Day", "t_life",
  "If propulsion stops, how long before the satellite re-enters?",
  "t = INT da/(da/dt) from h to 120 km",
  "vallado2013", "systems",
  [("h", "orbit_altitude"), ("bc", "aero_ballistic_coefficient"), ("t_inf", "env_exospheric_temperature")],
  [("integrate the decay rate down to the 120 km model base at fixed atmospheric conditions",
    "t", "Time",
    "orbit::lifetime_estimate(h, Length::from_km(120.0), bc.get(), 64, |z| env::mass_density(z, t_inf))")],
  0.0, 36500.0,
  "a lifetime cannot be negative",
  "above 100 years the 25-year debris rule is violated and the number's precision is irrelevant",
  tier="C",
  assumptions=[("Atmospheric conditions fixed at the current activity for the whole decay",
                "over a solar cycle the density at a given altitude changes by more than an order of magnitude, so this is an order-of-magnitude answer and is reported as one")],
  kpis=["kpi_service_lifetime"])
